cou/steps/plan.py

```diff
@@ -52,24 +54,162 @@
 logger = logging.getLogger(__name__)
 
 
-async def generate_plan(analysis_result: Analysis, backup_database: bool) -> UpgradePlan:
+def pre_plan_sanity_checks(args: CLIargs, analysis_result: Analysis) -> None:
+    """Pre checks to generate the upgrade plan.
+
+    :param args: CLI arguments
+    :type args: Namespace
+    :param analysis_result: Analysis result.
+    :type analysis_result: Analysis
+    """
+    is_valid_openstack_cloud(analysis_result)
+    is_supported_series(analysis_result)
+    is_highest_release_achieved(analysis_result)
+
+    if args.is_data_plane_command:
+        is_data_plane_ready_to_upgrade(analysis_result)
+
+
+def is_valid_openstack_cloud(analysis_result: Analysis) -> None:
+    """Check if the model passed is a valid OpenStack cloud.
+
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :raises NoTargetError: When cannot determine the current OS release
+        or Ubuntu series.
+    """
+    if not analysis_result.current_cloud_os_release:
+        raise NoTargetError(
+            "Cannot determine the current OS release in the cloud. "
+            "Is this a valid OpenStack cloud?"
+        )
+
+    if not analysis_result.current_cloud_series:
+        raise NoTargetError(
+            "Cannot determine the current Ubuntu series in the cloud. "
+            "Is this a valid OpenStack cloud?"
+        )
+
+
+def is_supported_series(analysis_result: Analysis) -> None:
+    """Check the Ubuntu series of the cloud to see if it is supported.
+
+    :param analysis_result: Analysis result.
+    :type analysis_result: Analysis
+    :raises OutOfSupportRange: When series is not supported.
+    """
+    supporting_lts_series = ", ".join(LTS_TO_OS_RELEASE)
+    # series already checked at is_valid_openstack_cloud
+    if (
+        current_series := analysis_result.current_cloud_series
+    ) and current_series not in LTS_TO_OS_RELEASE:
+        raise OutOfSupportRange(
+            f"Cloud series '{current_series}' is not a Ubuntu LTS series supported by COU. "
+            f"The supporting series are: {supporting_lts_series}"
+        )
+
+
+def is_highest_release_achieved(analysis_result: Analysis) -> None:
+    """Check if the highest OpenStack release is reached by the ubuntu series.
+
+    :param analysis_result: Analysis result.
+    :type analysis_result: Analysis
+    :raises HighestReleaseAchieved: When the OpenStack release is the last supported by the series.
+    """
+    if (
+        (current_os_release := analysis_result.current_cloud_os_release)
+        and (current_series := analysis_result.current_cloud_series)
+        and str(current_os_release) == LTS_TO_OS_RELEASE[current_series][-1]
+    ):
```

Comment:
Same here, what is the benefit doing it like this? Why not simply
```suggestion
    if str(analysis_result.current_cloud_os_release) == LTS_TO_OS_RELEASE[analysis_result.current_cloud_series][-1]:
```
or maybe even better would be:
```python
last_supported = LTS_TO_OS_RELEASE.get(analysis_result.current_cloud_series, [])[-1]
if str(analysis_result.current_cloud_os_release) == last_supported:
```
Like this, we show that we do not care if `current_cloud_series` exists in `LTS_TO_OS_RELEASE`

Reply:
mypy is a nightmare when we have `Optional[str]`. This code results in:
`error: Argument 1 to "get" of "dict" has incompatible type "str | None"; expected "str"  [arg-type]` that is why sometimes I prefer to use empty string instead of `None`.

We can disable the [type-checker](https://github.com/canonical/charmed-openstack-upgrader/pull/215/files#diff-abbb76655f6d778c97d8958cb0158d4fcc2f87854044f43695bcc788b76489e5R112) because `current_os_release` and `current_series` are already checked in another sanity check, however you seem to not like the approach.

Reply:
No, I do not like much the ignores, because when we are using mypy we rely on it, and when we disable it randomly it can cause an issue in the future. 

```python
series = analysis_result.current_cloud_series or ""
last_supported = LTS_TO_OS_RELEASE.get(series, [])[-1]
```
Or maybe we just need helper function to actually get `last_supported_series`?